xmake: 2.9.4 -> 2.9.5

rennsax opened the pull request:
## Description of changes

Xmake is actually broken in Nixpkgs. See #342966.

I'm surprised that this simple derivation "just works" - on my MacBook 2021 (aarch64-darwin), on my NixOS virtual machine (aarch64-linux) and on my lab's server (x86_64-linux).

However I'm not sure if the derivation is generic enough. Maybe more tests needed.

Here is the behavior of xmake:
```console
$ xmake create -l c -P ./hello
create hello ...
  [+]: xmake.lua
  [+]: src/main.c
  [+]: .gitignore
create ok!
$ cd hello
$ xmake
checking for platform ... linux
checking for architecture ... x86_64
[ 50%]: cache compiling.release src/main.c
[ 75%]: linking.release hello
[100%]: build ok, spent 1.287s
$ xmake --version
xmake v2.9.5+19800101, A cross-platform build utility based on Lua
Copyright (C) 2015-present Ruki Wang, tboox.org, xmake.io
                         _
    __  ___ __  __  __ _| | ______
    \ \/ / |  \/  |/ _  | |/ / __ \
     >  <  | \__/ | /_| |   <  ___/
    /_/\_\_|_|  |_|\__ \|_|\_\____|
                         by ruki, xmake.io

    👉  Manual: https://xmake.io/#/getting_started
    🙏  Donate: https://xmake.io/#/sponsor
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rennsax committed:
xmake: 2.9.4 -> 2.9.5

rennsax commented:
ping @wineee, the maintainer of xmake for reviewing

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,30 @@
+    patchShebangs --build configure
+  '';
 
-  meta = with lib; {
```
why remove mate?

rennsax replied:
Just a draft for concept proof. I'll add it later, apologize!

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,30 @@
+  stdenv,
+  fetchFromGitHub,
+  CoreServices,
+  ...
```
```suggestion

```

wineee replied:
Still a issue, there should be no useless input parameters

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,30 @@
-  ];
-
-  buildInputs = [
-    lua
```
Whenever possible, we should use build dependencies provided by nixpkgs instead of project‘s submodules.

rennsax replied:
Emmmm, that's an important point.

wineee replied:
tbox and xmake-core-sv are only used by xmake, so they should be upgraded to the versions required by xmake. If there are important bug fixes, you can also temporarily upgrade them to the unstable version.

rennsax replied:
> Whenever possible, we should use build dependencies provided by nixpkgs instead of project‘s submodules.

Then at least we need to build 7 extra packages for (and, only for) xmake, according to its [.gitsubmodule](https://github.com/xmake-io/xmake/blob/dev/.gitmodules)? Many dependencies of xmake are just forks of upstream projects with some specialized patches.

wineee replied:
The upstream author said that forking is only for submodules to facilitate git submodules management and fix it to a specific commit. Usually, no code of third-party dependencies will be changed.

https://github.com/xmake-io/xmake/discussions/3107

Atemu commented:
@wineee can I leave this one to you? You appear to have experience with this package.

rennsax commented:
@wineee If you DO NOT want to use the submodules in xmake repo to build it (just like the author of xmake did for Homebrew), please at least add `broken = true` to `xmake`. They are not well-tested and only available on your environment. And you should not add `xmake-core-sv` to Nixpkgs, because there is already `libsv`. Leave them in your NUR.

rennsax commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,29 @@
-    tbox
-    xmake-core-sv
-  ];
+  patches = [ ./0001-Fix-platform-compatibility-for-_os_find-and-_os_date.patch ];
```
This line can be removed after upgrading to 2.9.6. See https://github.com/xmake-io/xmake/pull/5655

wineee replied:
Use `fetchpatch`  
https://github.com/NixOS/nixpkgs/blob/master/doc/build-helpers/fetchers.chapter.md

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,29 @@
     homepage = "https://xmake.io";
     license = licenses.asl20;
-    platforms = lua.meta.platforms;
     maintainers = with maintainers; [ rewine ];
```
You can add yourself to maintainers if you want，I originally packaged xmake to build [mogan](https://github.com/NixOS/nixpkgs/issues/270803
), but the new version of mogan is more complicated, so I finally gave up

It would be good if there is an xmake user who maintains this package

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,29 @@
   pname = "xmake";
-  version = "2.9.4";
-
+  version = "v2.9.5";
```
```suggestion
  version = "2.9.5";
```
The 'v' prefix should not be in the `version`

rennsax replied:
Which version are you reviewing?

https://github.com/rennsax/nixpkgs/blob/1fe3e73b85eb5a16e2111dbf78b621439f0b1f2b/pkgs/development/tools/build-managers/xmake/default.nix#L1-L5

wineee commented on pkgs/development/tools/build-managers/xmake/default.nix:
```diff
@@ -1,48 +1,29 @@
-  configureFlags = [ "--external=y" ];
+  buildInputs =
+    [
+    ]
```
Remove empty brackets

rennsax committed:
xmake: add rennsax as the maintainer

wineee commented:
Please rebase and change commit message to:  `xmake: 2.9.4 -> 2.9.5`

rennsax commented:
Done. Thanks for reviewing.

wineee commented:
Build failed  on x86_64-darwin:

```
In file included from core/src/tbox/tbox/src/tbox/platform/file.c:41:
core/src/tbox/tbox/src/tbox/platform/posix/file.c:697:34: error: use of undeclared identifier 'UTIME_OMIT'
            else ts[0].tv_nsec = UTIME_OMIT;
                                 ^
core/src/tbox/tbox/src/tbox/platform/posix/file.c:699:34: error: use of undeclared identifier 'UTIME_OMIT'
            else ts[1].tv_nsec = UTIME_OMIT;
                                 ^
core/src/tbox/tbox/src/tbox/platform/posix/file.c:700:19: error: call to undeclared function 'utimensat'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
            ok = !utimensat(AT_FDCWD, path, ts, 0);
                  ^
core/src/tbox/tbox/src/tbox/platform/posix/file.c:715:38: error: use of undeclared identifier 'UTIME_OMIT'
                else ts[0].tv_nsec = UTIME_OMIT;
                                     ^
core/src/tbox/tbox/src/tbox/platform/posix/file.c:717:38: error: use of undeclared identifier 'UTIME_OMIT'
                else ts[1].tv_nsec = UTIME_OMIT;
                                     ^
core/src/tbox/tbox/src/tbox/platform/posix/file.c:718:23: error: call to undeclared function 'futimens'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                ok = !futimens(tb_file2fd(file), ts);
                      ^
6 errors generated.
```

https://logs.ofborg.org/?key=nixos/nixpkgs.343185&attempt_id=e2479983-f76a-4d28-a737-c1bc19a2ce6b

rennsax commented:
> Build failed on x86_64-darwin:
> 
> ```
> In file included from core/src/tbox/tbox/src/tbox/platform/file.c:41:
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:697:34: error: use of undeclared identifier 'UTIME_OMIT'
>             else ts[0].tv_nsec = UTIME_OMIT;
>                                  ^
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:699:34: error: use of undeclared identifier 'UTIME_OMIT'
>             else ts[1].tv_nsec = UTIME_OMIT;
>                                  ^
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:700:19: error: call to undeclared function 'utimensat'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>             ok = !utimensat(AT_FDCWD, path, ts, 0);
>                   ^
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:715:38: error: use of undeclared identifier 'UTIME_OMIT'
>                 else ts[0].tv_nsec = UTIME_OMIT;
>                                      ^
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:717:38: error: use of undeclared identifier 'UTIME_OMIT'
>                 else ts[1].tv_nsec = UTIME_OMIT;
>                                      ^
> core/src/tbox/tbox/src/tbox/platform/posix/file.c:718:23: error: call to undeclared function 'futimens'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>                 ok = !futimens(tb_file2fd(file), ts);
>                       ^
> 6 errors generated.
> ```
> 
> [logs.ofborg.org?key=nixos/nixpkgs.343185&attempt_id=e2479983-f76a-4d28-a737-c1bc19a2ce6b](https://logs.ofborg.org/?key=nixos/nixpkgs.343185&attempt_id=e2479983-f76a-4d28-a737-c1bc19a2ce6b)

Surprising 🙁 I'll test it later.

rennsax commented:
Find the reason and a related post: https://github.com/NixOS/nixpkgs/pull/338695

`utimensat` requires macOS SDK >= 10.13 but Nixpkgs will stick to 10.12.2 for `x86_64-darwin` until the release of 25.05.

wineee commented:
Add a new commit: `xmake:  fix buildon x86_64 darwin`

Also should use same sdk for CoreServices:
```diff
--     inherit (darwin.apple_sdk.frameworks) CoreServices;
++     inherit (darwin.apple_sdk_11_0.frameworks) CoreServices;
```

rennsax committed:
xmake: fix build on x86_64-darwin

xmake requires macOS SDK >= 10.13.0, but Nixpkgs sticks to macOS 10.12.0 for
x86_64-darwin now. Use 11.0 SDK as recommended by the doc.

rennsax commented:
Thanks
